File: src/lib/utils/rate-limiter.ts
export class RateLimiter {
  private timestamps: number[] = [];
  private maxRequests: number;
  private windowMs: number;

  constructor(maxRequests: number, windowMs: number) {
    this.maxRequests = maxRequests;
    this.windowMs = windowMs;
  }

  async waitForSlot(): Promise<void> {
    const now = Date.now();
    this.timestamps = this.timestamps.filter(t => now - t < this.windowMs);

    if (this.timestamps.length >= this.maxRequests) {
      const oldestInWindow = this.timestamps[0];
      const waitTime = this.windowMs - (now - oldestInWindow) + 100;
      await new Promise(resolve => setTimeout(resolve, waitTime));
      return this.waitForSlot();
    }

    this.timestamps.push(now);
  }
}

export const searchLimiter = new RateLimiter(2, 60000);
export const twitterLimiter = new RateLimiter(5, 60000);
export const aiLimiter = new RateLimiter(3, 10000);

File: agent-skills/hot-monitor/scripts/rate_limiter.py
"""Ported from src/lib/utils/rate-limiter.ts"""

from __future__ import annotations

import asyncio
import time


class RateLimiter:
    def __init__(self, max_requests: int, window_ms: int) -> None:
        self.max_requests = max_requests
        self.window_ms = window_ms
        self.timestamps: list[float] = []

    async def wait_for_slot(self) -> None:
        while True:
            now = time.time() * 1000
            self.timestamps = [t for t in self.timestamps if now - t < self.window_ms]
            if len(self.timestamps) < self.max_requests:
                self.timestamps.append(now)
                return
            oldest = self.timestamps[0]
            wait_ms = self.window_ms - (now - oldest) + 100
            await asyncio.sleep(wait_ms / 1000)


search_limiter = RateLimiter(2, 60_000)
twitter_limiter = RateLimiter(5, 60_000)
